### ISSUE TITLE
Agglomerating or hierarchical propagating tags

### ISSUE THREAD

**Issue by chambm** [galaxyproject/galaxy#4212]
I tried out @dannon's fix for adding workflow tags to dataset tags and it made me realize that the way dataset tags are presented in the history pane makes them pretty useless for analyses where many samples are combined. If I give each sample its own tag, then a combined dataset would have tags from all samples. In the current presentation, that would be unwieldy at best with a lot of samples. Collapsing individual dataset tags under a single tag header in the combined dataset could solve the problem. I can think of two ways to address this:

1. Add a new tag type, agglomerating, with a special prefix (e.g. `&sampleA`), that will trigger this collapsing behavior (and also propagate).

2. Add support for an optional hierarchy (or at least one-level parent category) to the propagating tags. This is a much more generalized approach and it would allow for multiple collapsing tags. For example, I can have 100 dataset tags like:
`#samples/1`
`#samples/2`
...
`#samples/100`

When looking at a dataset with just an individual tag, it would look like `sample/1`, but when looking at a dataset combining all 100 samples, it would show one tag `100 samples` which, if clicked on in the history pane, would show all the unique subtags.

As for generalizability, suppose I also wanted to analyze those samples with two methods for achieving about the same result (e.g. HISAT vs. STAR) which could considered independently or combined for even more sensitivity. I could have method tags like:
`#methods/HISAT`
`#methods/STAR`
When looking at datasets analyzed with just one method, it would show `method/HISAT` (or maybe just `HISAT`), but when looking at a dataset combining both methods, the tag would show up as `2 methods`, and clicking on the tag would show the unique subtags.

After writing this out, it seems option 2 would be both easier to implement and more generalizable. It's just a change to the presentation layer to support a delimiter to group tags to the left of the delimiter, right? To begin with 1 level of hierarchy would probably be fine.

**Comment by chambm** [comment]
Any feedback on this idea?

**Comment by mvdbeek** [comment]
Totally biased here, but I don't think tags are meant for samples, I use them to track groups of parallel execution (i.e Tumor/Normal pairs). Sample tracking should be done in collections, where each element is a sample IMO.

**Comment by mvdbeek** [comment]
> then a combined dataset would have tags from all samples

Specific tags can now be removed in the workflow editor as a PostJobAction, i.e if you subtract `Normal` from `Tumor` you don't want to keep the `Normal` tag. Maybe that helps for what you want to do ?

**Comment by mvdbeek** [comment]
So basically I use tags in my workflows/histories to avoid nested lists, where the highest level would be the group, and the lowest level the samples in the group. But nested lists are a little trickier to think about and harder to construct, but if you have many groups they would probably be cleaner.

**Comment by chambm** [comment]
Using tags for samples is the safest way to ensure that some workflow step doesn't clobber the sample reference by changing the name in a bad way. Well, assuming the step doesn't also clobber the tags. ;)

**Comment by mvdbeek** [comment]
Collections are just as safe if not safer because they are immutable.

**Comment by chambm** [comment]
It's true that collection element names are very inflexible, so that does avoid the problem. Actually it's the names of the collections themselves that get screwed up when trying to use PostJobActions to rename the outputs...

**Comment by mvdbeek** [comment]
True, I avoid the renaming operations altogether. We should either augment that for collections are remove this for collection steps.

**Comment by chambm** [comment]
What about searching for all the files related to a sample? If the rename operations don't preserve the sample name in the actual datasets, and there's no tags, you're pretty much SOL, right?

**Comment by mvdbeek** [comment]
That is something I have never done in practice, given that I usually know in what state (=collection) I'm looking for my sample, and if I need to find the precursor I can walk up the chain. But we could augment the search to search for element identifiers

**Comment by chambm** [comment]
Yes, that would be a simpler solution I guess. If this is best practice, then there should be some guidance with propagating tags to use them only for low-cardinality sets. :)

**Comment by mvdbeek** [comment]
But I don't think we have a best-practice there, that's why I think this is an interesting discussion.

**Comment by dannon** [comment]
Definitely an interesting discussion.  I'm interested to see what more users think would be the friendliest thing to do here is.  Currently I think I lean towards @mvdbeek's suggestion of just using Collections for this instead.
